src/vs/workbench/contrib/chat/browser/chatPasteProviders.ts

```diff
@@ -12,26 +12,96 @@ import { ITextModel } from '../../../../editor/common/model.js';
 import { ILanguageFeaturesService } from '../../../../editor/common/services/languageFeatures.js';
 import { Disposable } from '../../../../base/common/lifecycle.js';
 import { ChatInputPart } from './chatInputPart.js';
+import { IClipboardService } from '../../../../platform/clipboard/common/clipboardService.js';
+import { IChatWidgetService } from './chat.js';
+import { Codicon } from '../../../../base/common/codicons.js';
+import { localize } from '../../../../nls.js';
+import { IChatRequestVariableEntry } from '../common/chatModel.js';
+import { IConfigurationService } from '../../../../platform/configuration/common/configuration.js';
+import { ChatContextAttachments } from './contrib/chatContextAttachments.js';
 
 
 export class PasteImageProvider implements DocumentPasteEditProvider {
 
 	public readonly kind = new HierarchicalKind('image');
 
 	public readonly pasteMimeTypes = ['image/*'];
+	constructor(
+		private readonly clipboardService: IClipboardService,
+		private readonly chatWidgetService: IChatWidgetService,
+		private readonly configurationService: IConfigurationService
+	) { }
 
 	async provideDocumentPasteEdits(_model: ITextModel, _ranges: readonly IRange[], dataTransfer: IReadonlyVSDataTransfer, context: DocumentPasteContext, token: CancellationToken): Promise<DocumentPasteEditsSession | undefined> {
-		// TODO: @justschen setup image data types
-		// const entry = dataTransfer.get('image/png') || dataTransfer.get('image/jpeg') || dataTransfer.get('image/gif') || dataTransfer.get('image/webp');
+		if (!this.configurationService.getValue<boolean>('chat.experimental.imageAttachments')) {
+			return;
+		}
+
+		const currClipboard = await this.clipboardService.readImage();
+
+		if (!currClipboard || !isImage(currClipboard)) {
+			return;
+		}
+		const imageContext = await getImageAttachContext(currClipboard);
+		if (!imageContext) {
+			return;
+		}
+		const widget = this.chatWidgetService.getWidgetByInputUri(_model.uri);
+		if (!widget) {
+			return;
+		}
+
+		const attachContribution = widget.getContrib<ChatContextAttachments>(ChatContextAttachments.ID);
+		attachContribution?.setContext(false, imageContext);
```

Comment:
`provideDocumentPasteEdits` should not have any side effects. Maybe ok as a temporary workaround for now but let's look into allowing workspace edits to add attachments to chat. That way undo/redo will work nicely too